scripts/yetong01_kinematic_trajectory_planning.cpp

```diff
@@ -0,0 +1,345 @@
+Values optimize_dynamics_graph(const NonlinearFactorGraph& constraints_graph,
+                               const NonlinearFactorGraph& costs,
+                               const Values& init_values) {
+  NonlinearFactorGraph graph;
+  graph.add(constraints_graph);
+  graph.add(costs);
+
+  LevenbergMarquardtParams params;
+  params.setVerbosityLM("SUMMARY");
+  LevenbergMarquardtOptimizer optimizer(graph, init_values, params);
+  gttic_(dynamic_factor_graph);
+  auto result = optimizer.optimize();
+  gttoc_(dynamic_factor_graph);
+
+  size_t graph_dim = 0;
+  for (const auto& factor : graph) {
+    graph_dim += factor->dim();
+  }
+  std::cout << "dimension: " << graph_dim << " x " << init_values.dim() << "\n";
+  return result;
+}
+
+/** Kinematic trajectory optimization using constraint manifold. */
+Values optimize_constraint_manifold(
+    const NonlinearFactorGraph& constraints_graph,
+    const NonlinearFactorGraph& costs, const Values& init_values) {
+  auto constraints = ConstraintsFromGraph(constraints_graph);
+
+  LevenbergMarquardtParams params;
+  params.setVerbosityLM("SUMMARY");
+  // params.minModelFidelity = 0.1;
+  std::cout << "building optimizer\n";
+  ManifoldOptimizerType1 optimizer(costs, constraints, init_values, params);
+  std::cout << "optimize\n";
+  gttic_(constraint_manifold);
+  auto result = optimizer.optimize();
+  gttoc_(constraint_manifold);
+
+  auto problem_dim = optimizer.problem_dimension();
+  std::cout << "dimension: " << problem_dim.first << " x " << problem_dim.second
+            << "\n";
+  return result;
+}
+
+/** Kinematic trajectory optimization using manually defined serial chain
+ * manifold. */
+Values optimize_serial_chain_manifold(const NonlinearFactorGraph& costs,
+                                      const Values& init_values) {
+  LevenbergMarquardtParams params;
+  params.setVerbosityLM("SUMMARY");
+  size_t graph_dim = 0;
+  for (const auto& factor : costs) {
+    graph_dim += factor->dim();
+  }
+  std::cout << "dimension: " << graph_dim << " x " << init_values.dim() << "\n";
+  LevenbergMarquardtOptimizer optimizer(costs, init_values, params);
+  gttic_(serial_chain_manifold);
+  auto result = optimizer.optimize();
+  gttoc_(serial_chain_manifold);
+  return result;
+}
+
+/** Factor graph of all kinematic constraints. */
+NonlinearFactorGraph get_constraints_graph() {
+  NonlinearFactorGraph constraints_graph;
+  // kinematics constraints at each step
+  auto graph_builder = DynamicsGraph();
+  for (size_t k = 0; k <= num_steps; k++) {
+    constraints_graph.add(graph_builder.qFactors(robot, k));
+  }
+  // constraints_graph.print("constraints_graph", GTDKeyFormatter);
+  // // prior constraints at first step
+  size_t k0 = 0;
+  for (const auto& joint : robot.joints()) {
+    constraints_graph.addPrior(JointAngleKey(joint->id(), k0), 0.0,
+                               graph_builder.opt().prior_q_cost_model);
+  }
+  return constraints_graph;
+}
+
+/** Cost. */
```

Comment:
Sorry but this needs to be much better. What costs are we talking about? What is being computed and what should be expected?